### PR TITLE
Use ReflectionBasedSerializer for all serializers if any of mappings are soap mappings

### PR DESCRIPTION
**Pull request by huanwu**
When creating serializers for multiple mappings, if any of mappings are soap mapping, it will use ReflectionBasedSerializer and won't generate the serializer. However during serialization, for non soap mapping, it will try to use the serializer for serialization, which doesn't exist. The fix is to force all serializers using ReflectionBasedSerializer if there's any soap mapping.

#25298 
@zhenlan @mconnew @yujayee @Lxiamail

### CODE REVIEW

**Comment by huanwu** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by huanwu** [comment]
@dotnet-bot test NETFX x86 Release Build

**Comment by huanwu** [comment]
@dotnet-bot test Tizen armel Debug Build

**Comment by zhenlan** [comment]
`Deserialize` calls `ShouldUseReflectionBasedSerialization` too. Will it have the same problem?
https://github.com/dotnet/corefx/pull/27803/files#diff-b5f806042e753aadb89073779fa4ee17R591

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs [hunk 1]
```diff
@@ -447,7 +448,7 @@ public void Serialize(XmlWriter xmlWriter, object o, XmlSerializerNamespaces nam
                     }
                     SerializePrimitive(xmlWriter, o, namespaces);
                 }
-                else if (ShouldUseReflectionBasedSerialization(_mapping))
+                else if (ShouldUseReflectionBasedSerialization(_mapping) || _isReflectionBasedSerializer)
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: I am wondering if it is still necessary to check `ShouldUseReflectionBasedSerialization` here.

Comment (huanwu) [reply src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: No, we cannot remove it.  _isReflectionBasedSerializer is only being set in FromMappings in this case. When to create a serializer from a single soap mapping, it won't set this value, and it need use this method to check if it is a soap mapping. In addition, this method also check if the SerializationMode is reflectiononly or not.

**Comment by huanwu** [comment]
@dotnet-bot test OSX x64 Debug Build

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs (hunk 1 above)
Comment (zhenlan) [reply src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: Would it be feasible to consolidate all the logic of using reflection based serialization in one place?

Comment (huanwu) [reply src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs]: It is better to seperate them. ShouldUseReflectionBasedSerialization method is a static method but _isReflectionBasedSerializer is not static.

**Comment by huanwu** [comment]
@zhenlan The test need be added into wcf. So I need wait until the wcf take the new assembly.

**Comment by zhenlan** [comment]
> The test need be added into wcf. So I need wait until the wcf take the new assembly.

Sounds good. Thanks, @huanwu!

**Pull request merged**